packages/@aws-cdk/aws-ecs/lib/container-definition.ts

```diff
@@ -551,23 +564,35 @@ export class ContainerDefinition extends cdk.Construct {
       workingDirectory: this.props.workingDirectory,
       logConfiguration: this.logDriverConfig,
       environment: this.props.environment && renderKV(this.props.environment, 'name', 'value'),
-      secrets: this.props.secrets && Object.entries(this.props.secrets)
-        .map(([k, v]) => {
-          if (taskDefinition) {
-            v.grantRead(taskDefinition.obtainExecutionRole());
-          }
-          return {
-            name: k,
-            valueFrom: v.arn
-          };
-        }),
+      secrets: this.renderSecrets(),
       extraHosts: this.props.extraHosts && renderKV(this.props.extraHosts, 'hostname', 'ipAddress'),
       healthCheck: this.props.healthCheck && renderHealthCheck(this.props.healthCheck),
       links: cdk.Lazy.listValue({ produce: () => this.links }, { omitEmpty: true }),
       linuxParameters: this.linuxParameters && this.linuxParameters.renderLinuxParameters(),
       resourceRequirements: (this.props.gpuCount !== undefined) ? renderResourceRequirements(this.props.gpuCount) : undefined,
     };
   }
+
+  private renderSecrets(): CfnTaskDefinition.SecretProperty[] | undefined {
+    if (!this.props.secrets) {
+      return undefined;
+    }
+
+    const secrets: CfnTaskDefinition.SecretProperty[] = [];
+
+    for (const [k, v] of Object.entries(this.props.secrets)) {
+      if (this.taskDefinition.isFargateCompatible && v.hasField) {
+        throw new Error(`Cannot specify secret JSON field for a task using the FARGATE launch type: '${k}' in container '${this.node.id}'`);
```

Comment:
Do this here or in the constructor?

Reply:
This would probably make more sense to be in the constructor.